### PR TITLE
[WIP] Test using jpeg compression 

### PR DESCRIPTION
**Pull request by VibhuJawa**
## Description

Benchmarking Results (Speed up is 9.23x) :

```
(triton_bare_bones) vjawa@viking-prod-206:~/nv-ingest/src/nv_ingest/util/image_processing$ python3 benchmark_transforms.py 
Creating test image...
Test image shape: (1024, 1024, 3), dtype: uint8

Running timing benchmarks...
--------------------------------------------------
Old implementation (PIL-based):
Average time: 124.11 ms
Output size: 4103.09 KB

New implementation (torchvision-based):
Average time: 13.45 ms
Output size: 1599.02 KB

Speedup: 9.23x

Running memory benchmarks...
--------------------------------------------------
Old implementation memory usage:
Filename: /home/nfs/vjawa/nv-ingest/src/nv_ingest/util/image_processing/benchmark_transforms.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
   135    547.6 MiB    547.6 MiB           1   @memory_profiler.profile
   136                                         def run_memory_benchmark(func, array):
   137                                             """Run memory benchmark for a given implementation."""
   138    550.0 MiB      2.5 MiB           1       return func(array)

New implementation memory usage:
Filename: /home/nfs/vjawa/nv-ingest/src/nv_ingest/util/image_processing/benchmark_transforms.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
   135    550.0 MiB    550.0 MiB           1   @memory_profiler.profile
   136                                         def run_memory_benchmark(func, array):
   137                                             """Run memory benchmark for a given implementation."""
   138    550.0 MiB      0.0 MiB           1       return func(array)
```

<details>
<summary>Benchmarking Script</summary>

```python
import time
import numpy as np
import torch
import memory_profiler
from PIL import Image
import io
import base64
import torch
import base64
from torchvision.io import encode_jpeg



def base64frombytes(bytes_input, encoding="utf-8"):
    """
    Function to bytes to base64 string.

    Parameters
    ----------
    bytes_input : bytes
        Raw bytes of object.

    Returns
    -------
    base64
        base64 encoded string to store bytes in cuDF.
    """

    return base64.b64encode(bytes_input).decode(encoding)

def torchvision_numpy_to_base64(array: np.ndarray, quality: int = 95) -> str:
    """
    Converts a NumPy array representing an image to a base64-encoded JPEG string using torchvision.

    The function takes a NumPy array, converts it to a torch tensor, and then encodes
    the image as a JPEG in a base64 string format. The input array is expected to be in
    a format that can be converted to a valid image, such as having a shape of (H, W, C)
    where C is the number of channels (e.g., 3 for RGB).

    Parameters
    ----------
    array : np.ndarray
        The input image as a NumPy array. Must have a shape compatible with image data.
        Can be (H, W) for grayscale or (H, W, C) where C∈{1,3,4}. 4-channel images lose alpha.
    quality : int, optional
        JPEG quality (1-100), by default 95

    Returns
    -------
    str
        The base64-encoded string representation of the input NumPy array as a JPEG image.

    Raises
    ------
    ValueError
        If the input array cannot be converted into a valid image format.
    RuntimeError
        If there is an issue during the image conversion or base64 encoding process.

    Examples
    --------
    >>> array = np.random.randint(0, 255, (100, 100, 3), dtype=np.uint8)
    >>> encoded_str = numpy_to_base64(array)
    >>> isinstance(encoded_str, str)
    True
    """

    # Handle grayscale images with shape (H, W, 1)
    if array.ndim == 3 and array.shape[2] == 1:
        array = np.squeeze(array, axis=2)

    # Convert to torch tensor with appropriate shape
    if array.ndim == 2:  # grayscale
        tensor = torch.from_numpy(array).unsqueeze(0)  # (1, H, W)
    elif array.ndim == 3:
        if array.shape[2] == 4:  # drop alpha channel
            array = array[..., :3]
        tensor = torch.from_numpy(array).permute(2, 0, 1)  # (C, H, W)
    else:
        raise ValueError(f"Expected (H,W) or (H,W,C); got {array.shape}")

    # Ensure uint8 dtype
    if tensor.dtype != torch.uint8:
        tensor = tensor.to(torch.uint8)

    try:
        # Encode to JPEG using torchvision
        jpeg_bytes_tensor = encode_jpeg(
            tensor.contiguous(),
            quality=quality,
        )
        jpeg_bytes = jpeg_bytes_tensor.cpu().numpy().tobytes()
    except Exception as e:
        raise RuntimeError(f"Failed to encode JPEG with torchvision: {e}")

    # Convert to base64
    return base64frombytes(jpeg_bytes)

    

def pil_numpy_to_base64(array: np.ndarray) -> str:
    """Original PIL-based implementation for benchmarking."""
    if array.ndim == 3 and array.shape[2] == 1:
        array = np.squeeze(array, axis=2)

    try:
        pil_image = Image.fromarray(array.astype(np.uint8))
    except Exception as e:
        raise ValueError(f"Failed to convert NumPy array to image: {e}")

    try:
        with io.BytesIO() as buffer:
            pil_image.save(buffer, format="PNG")
            base64_img = base64frombytes(buffer.getvalue())
    except Exception as e:
        raise RuntimeError(f"Failed to encode image to base64: {e}")

    return base64_img

def benchmark_implementation(func, array, num_runs=100):
    """Run timing benchmark for a given implementation."""
    # Warmup
    for _ in range(10):
        func(array)
    
    # Actual timing
    start_time = time.perf_counter()
    for _ in range(num_runs):
        result = func(array)
    end_time = time.perf_counter()
    
    avg_time = (end_time - start_time) / num_runs
    return avg_time, len(result)

@memory_profiler.profile
def run_memory_benchmark(func, array):
    """Run memory benchmark for a given implementation."""
    return func(array)

def main():
    # Create test image
    print("Creating test image...")
    test_image = np.random.randint(0, 255, (1024, 1024, 3), dtype=np.uint8)
    print(f"Test image shape: {test_image.shape}, dtype: {test_image.dtype}")
    
    # Benchmark parameters
    num_runs = 100
    
    print("\nRunning timing benchmarks...")
    print("-" * 50)
    
    # Benchmark old implementation
    print("Old implementation (PIL-based):")
    old_time, old_size = benchmark_implementation(pil_numpy_to_base64, test_image, num_runs)
    print(f"Average time: {old_time*1000:.2f} ms")
    print(f"Output size: {old_size/1024:.2f} KB")
    
    # Benchmark new implementation
    print("\nNew implementation (torchvision-based):")
    new_time, new_size = benchmark_implementation(torchvision_numpy_to_base64, test_image, num_runs)
    print(f"Average time: {new_time*1000:.2f} ms")
    print(f"Output size: {new_size/1024:.2f} KB")
    
    # Calculate speedup
    speedup = old_time / new_time
    print(f"\nSpeedup: {speedup:.2f}x")
    
    # Memory benchmarks
    print("\nRunning memory benchmarks...")
    print("-" * 50)
    
    print("Old implementation memory usage:")
    old_memory = run_memory_benchmark(pil_numpy_to_base64, test_image)
    
    print("\nNew implementation memory usage:")
    new_memory = run_memory_benchmark(torchvision_numpy_to_base64, test_image)
    
    # Verify outputs are valid
    print("\nVerifying outputs...")
    try:
        base64.b64decode(pil_numpy_to_base64(test_image))
        print("Old implementation output is valid base64")
    except Exception as e:
        print(f"Old implementation output is invalid: {e}")
        
    try:
        base64.b64decode(torchvision_numpy_to_base64(test_image))
        print("New implementation output is valid base64")
    except Exception as e:
        print(f"New implementation output is invalid: {e}")

if __name__ == "__main__":
    main() 
```

</details>

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/nv-ingest/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] If adjusting docker-compose.yaml environment variables have you ensured those are mimicked in the Helm values.yaml file.

### CODE REVIEW

**Comment by randerzander** [comment]
run benchmarks

**Comment by randerzander** [comment]
Benchmark Results: :x:
PR number: 721

Notes: N/A

GPUs: NVIDIA H100 NVL

nv-ingest-cli check: :green-check-mark:
helm check: :x:
library mode check: :green-check-mark:
audio check: :green-check-mark:
pptx/docx/image/txt check: :green-check-mark:

Latency Test (bo20 PDFs) :green-check-mark::
20/20 docs, 496 pages in 71.34 extraction seconds: 6.95 pages/sec
79.99 e2e (includes indexing) seconds: 6.20 pages/sec

Throughput Test (bo767 PDFs) :green-check-mark::
767/767 docs in 92 minutes:
9.91 e2e pages/sec

Bo767 Multimodal Recall Accuracy:
Without re-ranker: (51) queries/sec:
recall @1: 0.565
recall @3: 0.773
recall @5: 0.830
recall @10: 0.881

With re-ranker: (2) queries/sec:
recall @1: 0.729
recall @3: 0.855
recall @5: 0.892
recall @10: 0.917

Earnings Multimodal Recall Accuracy:
Without re-ranker:
recall @1: 0.369
recall @3: 0.557
recall @5: 0.619
recall @10: 0.726

With re-ranker:
recall @1: 0.511
recall @3: 0.670
recall @5: 0.732
recall @10: 0.790

nv-ingest sha: e42c2fd3f9fedb152bd8624674d6cef4d038c5a8

Images:
riva-asr:latest
nemoretriever-graphic-elements-v1:1.3.0
nemoretriever-page-elements-v2:1.3.0
llama-3.2-nv-rerankqa-1b-v2:1.5.0
paddleocr:1.3.0
nemoretriever-table-structure-v1:1.3.0
nv-ingest:25.4.2
llama-3.2-nv-embedqa-1b-v2:1.6.0
dcgm-exporter:4.1.0-4.0.2-ubuntu22.04

**Comment by VibhuJawa** [comment]
run integration tests

**Comment by randerzander** [comment]
Integration Test Results: 
  GPUs: NVIDIA H100 NVL

  nv-ingest-cli check: :green-check-mark:
  helm check: :x:
  library mode check: :green-check-mark:
  audio check: :green-check-mark:
  pptx/docx/image/txt check: :green-check-mark:

  Latency Test (bo20 PDFs) :green-check-mark::
  20/20 docs, 496 pages in 70.67 extraction seconds: 7.02 pages/sec
  78.86 e2e (includes indexing) seconds: 6.29 pages/sec

**Comment by VibhuJawa** [comment]
run integration tests

**Comment by randerzander** [comment]
Integration Test Results: 
  GPUs: NVIDIA H100 NVL

  nv-ingest-cli check: :white_check_mark:
  helm check: :x:
  library mode check: :x:
  audio check: :x:
  pptx/docx/image/txt check: :white_check_mark:

  Latency Test (bo20 PDFs) :white_check_mark::
  20/20 docs, 496 pages in 78.55 extraction seconds: 6.31 pages/sec
  85.32 e2e (includes indexing) seconds: 5.81 pages/sec

  Docker Images:
  riva-asr:latest
paddleocr:1.3.0
nv-ingest:25.3.0
llama-3.2-nv-embedqa-1b-v2:1.6.0
nemoretriever-graphic-elements-v1:1.3.0
llama-3.2-nv-rerankqa-1b-v2:1.5.0
nemoretriever-table-structure-v1:1.3.0
nemoretriever-page-elements-v2:1.3.0
dcgm-exporter:4.1.0-4.0.2-ubuntu22.04

**Comment by VibhuJawa** [comment]
run integration tests

**Comment by VibhuJawa** [comment]
run benchmarks

File: api/src/nv_ingest_api/internal/extract/pdf/engines/nemoretriever.py [hunk 1]
```diff
@@ -217,21 +219,30 @@ def nemoretriever_parse_extractor(
...
+                p_img, _ = _convert_pdfium_page_to_numpy_for_parser(p)
+                y_img, y_pad = _convert_pdfium_page_to_numpy_for_yolox(p)
+                p.close()
+            return idx, p_img, y_img, y_pad
+
+        render_workers = min(cpu_count(), 8)
+        render_pool = ThreadPoolExecutor(max_workers=render_workers)
+        render_futs = [
+            render_pool.submit(_render_page, pdf_bytes, i) for i in range(page_count)
+        ]
```
Comment (VibhuJawa) [inline api/src/nv_ingest_api/internal/extract/pdf/engines/nemoretriever.py]: @jdye64 ,  I wonder if we can do something like this to first call render pages in parallel, so that our yolox model is not waiting around always ?

**Comment by VibhuJawa** [comment]
Closing in support of https://github.com/NVIDIA/nv-ingest/pull/798
